File: clients/quarkus/flowcatalyst-sdk/src/main/java/tech/flowcatalyst/sdk/client/resources/Roles.java
package tech.flowcatalyst.sdk.client.resources;

import com.fasterxml.jackson.core.type.TypeReference;
import tech.flowcatalyst.sdk.client.FlowCatalystClient;
import tech.flowcatalyst.sdk.dto.ListResult;
import tech.flowcatalyst.sdk.dto.Role;

import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Resource for managing roles.
 */
public class Roles {

    private final FlowCatalystClient client;

    public Roles(FlowCatalystClient client) {
        this.client = client;
    }

    /**
     * List all roles.
     */
    public ListResult<Role> list() {
        return list(Map.of());
    }

    /**
     * List roles with filters.
     */
    public ListResult<Role> list(Map<String, String> filters) {
        String query = filters.isEmpty() ? "" : "?" + buildQuery(filters);
        var response = client.request("GET", "/api/roles" + query, null,
            new TypeReference<Map<String, Object>>() {});

        @SuppressWarnings("unchecked")
        var items = (List<Map<String, Object>>) response.getOrDefault("items", List.of());
        var roles = items.stream()
            .map(this::mapRole)
            .toList();

        return ListResult.of(roles);
    }

    /**
     * Get a role by ID.
     */
    public Role get(String id) {
        var response = client.request("GET", "/api/roles/" + id, null,
            new TypeReference<Map<String, Object>>() {});
        return mapRole(response);
    }

    /**
     * Create a new role.
     */
    public Role create(CreateRoleRequest request) {
        var response = client.request("POST", "/api/roles", request,
            new TypeReference<Map<String, Object>>() {});
        return mapRole(response);
    }

    /**
     * Update a role.
     */
    public Role update(String id, UpdateRoleRequest request) {
        var response = client.request("PATCH", "/api/roles/" + id, request,
            new TypeReference<Map<String, Object>>() {});
        return mapRole(response);
    }

    /**
     * Delete a role.
     */
    public void delete(String id) {
        client.requestVoid("DELETE", "/api/roles/" + id, null);
    }

    /**
     * Sync roles for an application (SDK-managed roles).
     *
     * <p>This will create/update roles to match the provided list and optionally
     * remove roles not in the list.
     */
    public SyncResult sync(String applicationCode, List<SyncRoleDefinition> roles, boolean removeUnlisted) {
        var request = new SyncRolesRequest(applicationCode, roles, removeUnlisted);
        var response = client.request("POST", "/api/roles/sync", request,
            new TypeReference<Map<String, Object>>() {});

        @SuppressWarnings("unchecked")
        var created = (List<String>) response.getOrDefault("created", List.of());
        @SuppressWarnings("unchecked")
        var updated = (List<String>) response.getOrDefault("updated", List.of());
        @SuppressWarnings("unchecked")
        var deleted = (List<String>) response.getOrDefault("deleted", List.of());

        return new SyncResult(created, updated, deleted);
    }

    private Role mapRole(Map<String, Object> data) {
        return client.getObjectMapper().convertValue(data, Role.class);
    }

    private String buildQuery(Map<String, String> params) {
        return params.entrySet().stream()
            .map(e -> e.getKey() + "=" + e.getValue())
            .reduce((a, b) -> a + "&" + b)
            .orElse("");
    }

    // Request/Response DTOs

    public record CreateRoleRequest(
        String applicationId,
        String name,
        String displayName,
        String description,
        Set<String> permissions,
        boolean clientManaged
    ) {}

    public record UpdateRoleRequest(
        String displayName,
        String description,
        Set<String> permissions
    ) {}

    public record SyncRolesRequest(
        String applicationCode,
        List<SyncRoleDefinition> roles,
        boolean removeUnlisted
    ) {}

    public record SyncRoleDefinition(
        String name,
        String displayName,
        String description,
        Set<String> permissions
    ) {}

    public record SyncResult(
        List<String> created,
        List<String> updated,
        List<String> deleted
    ) {}
}

File: packages/platform-ui-vue/src/api/roles.ts
import { bffFetch } from './client';

export type RoleSource = 'CODE' | 'DATABASE' | 'SDK';

export interface Role {
  id: string;
  name: string;
  shortName: string;
  displayName: string;
  description?: string;
  permissions: string[];
  applicationCode: string;
  source: RoleSource;
  clientManaged: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface RoleListResponse {
  items: Role[];
  total: number;
}

export interface ApplicationOption {
  id: string;
  code: string;
  name: string;
}

export interface ApplicationOptionsResponse {
  options: ApplicationOption[];
}

export interface CreateRoleRequest {
  applicationCode: string;
  name: string;
  displayName?: string;
  description?: string;
  permissions?: string[];
  clientManaged?: boolean;
}

export interface UpdateRoleRequest {
  displayName?: string;
  description?: string;
  permissions?: string[];
  clientManaged?: boolean;
}

export interface RoleFilters {
  application?: string;
  source?: RoleSource;
}

export const rolesApi = {
  list(filters: RoleFilters = {}): Promise<RoleListResponse> {
    const params = new URLSearchParams();
    if (filters.application) params.set('application', filters.application);
    if (filters.source) params.set('source', filters.source);

    const query = params.toString();
    return bffFetch(`/roles${query ? `?${query}` : ''}`);
  },

  get(roleName: string): Promise<Role> {
    return bffFetch(`/roles/${encodeURIComponent(roleName)}`);
  },

  create(data: CreateRoleRequest): Promise<Role> {
    return bffFetch('/roles', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  update(roleName: string, data: UpdateRoleRequest): Promise<Role> {
    return bffFetch(`/roles/${encodeURIComponent(roleName)}`, {
      method: 'PUT',
      body: JSON.stringify(data),
    });
  },

  delete(roleName: string): Promise<void> {
    return bffFetch(`/roles/${encodeURIComponent(roleName)}`, { method: 'DELETE' });
  },

  // Filter options
  getApplications(): Promise<ApplicationOptionsResponse> {
    return bffFetch('/roles/filters/applications');
  },
};
